src/pg_copy.c

```diff
@@ -0,0 +1,574 @@
+	equalityExpr = MakeOpExpression(partitionColumn, BTEqualStrategyNumber);
+	rightOp = get_rightop((Expr *) equalityExpr);				 
+	Assert(IsA(rightOp, Const));
+	rightConst = (Const *) rightOp;
+	rightConst->constvalue = 0;
+	rightConst->constisnull = false;
+	rightConst->constbyval = get_typbyval(columnOid);
+	whereClauseList = list_make1(equalityExpr);
+
+	/* allocate column values and nulls arrays */
+	rel = heap_open(tableId,  AccessShareLock);
+	tupleDescriptor = RelationGetDescr(rel);
+	columnCount = tupleDescriptor->natts;
+	columnValues = palloc0(columnCount * sizeof(Datum));
+	columnNulls = palloc0(columnCount * sizeof(bool));
+
+	/*
+	 * We create a new memory context called tuple context, and read and write
+	 * each row's values within this memory context. After each read and write,
+	 * we reset the memory context. That way, we immediately release memory
+	 * allocated for each row, and don't bloat memory usage with large input
+	 * files.
+	 */
+	tupleContext = AllocSetContextCreate(CurrentMemoryContext,
+										 "COPY Row Memory Context",
+										 ALLOCSET_DEFAULT_MINSIZE,
+										 ALLOCSET_DEFAULT_INITSIZE,
+										 ALLOCSET_DEFAULT_MAXSIZE);
+			
+	/* 
+	 * Construct hash table used for shardId->Connection mapping.
+	 * We can not use connection cache from connection.c used by GteConnection because
+	 * we need to establish multiple connections with each nodes: one connection per shard
+	 */
+	shardToConn = CreateShardToConnectionHash();
+	/* init state to read from COPY data source */
+	copyState = BeginCopyFrom(rel, copyStatement->filename,
+							  copyStatement->is_program,
+							  copyStatement->attlist,
+							  copyStatement->options);
+				
+	if (copyState->binary) 
+	{ 
+		elog(ERROR, "Copy in binary mode is not currently supported");
+	}
+
+	PG_TRY();
+	{
+		/* Lock all shards in shared mode */
+		shardIntervalList = SortList(shardIntervalList, CompareTasksByShardId);
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+		}
+
+		while (true)
+		{
+			MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+			nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+			MemoryContextSwitchTo(oldContext);
+		
+			if (!nextRowFound)
+			{
+				MemoryContextReset(tupleContext);
+				break;
+			}
+			
+			/* write the row to the shard */
+			if (columnNulls[partitionColumn->varattno]) 
+			{ 
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			} 
+			rightConst->constvalue = columnValues[partitionColumn->varattno];
+			prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+				
+			foreach(shardIntervalCell, prunedList)
```

Comment:
A shardInterval represents a single shard (part of the table). If you copy into multiple shards, you're copying into the table multiple times. 

The reason PruneShardList might return multiple shards is that a) SELECT/UPDATE/DELETE queries could hit multiple shards, b) multiple shards could have overlapping intervals if users manually change the metadata. We only need to consider case b here. If multiple shards overlap, we should copy to one of them. For correctness, it does not matter which one, so you can just use linitial(prunedList) and remove the foreach loop. Copying to all replicas of the shard is handled at line 539.